enhancements/mcd-reboot-exceptions.md

```diff
@@ -0,0 +1,247 @@
+---
+title: mcd-reboot-dictionary
+authors:
+  - "@beekhof"
+reviewers:
+  - "@crawford"
+  - "@eparis"
+  - "@rphillips"
+  - "@runcom"
+  - "@yuqi-zhang"
+  - "@kikisdeliveryservice"
+  - "@sinnykumari"
+  - "@ericavonb"
+approvers:
+  - "@crawford"
+  - "@eparis"
+  - "@rphillips"
+  - "@runcom"
+creation-date: 2019-12-15
+last-updated: 2020-06-18
+status: provisional
+see-also:
+  - NA
+replaces:
+  - NA
+superseded-by:
+  - NA
+---
+
+# Domain specific minimisation of config driven cluster reboots
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Should the parts of the configuration that do and don't require reboot be
+  part of user facing documentation?
+
+## Summary
+
+A node's configuration is managed by the MCO.  When the configuation is changed
+by the system or an admin, the MCO and MCD write a new version of the file to
+disk and reboot the system to ensure it is applied.
+
+This proposal is to allow drain and reboot to be optional and independant
+actions, to implement the ability to restart a systemd service, and to add a
+function to MCD that compares two MachineConfigs and can decide which actions
+are needed to apply it.
+
+
+## Motivation
+
+The current MCO mechanism for applying config changes, is to reboot the machine.
+
+While it is expected and normal that any cluster node can reboot at any time,
+clusters have a practical limit to the number of full cluster reboots they can
+achieve in a given time period.
+
+This means we need to weigh the correctness and simplicity of using full cluster
+reboots, against their frequency and user impact.
+
+In environments where the limit is generally lower, such as bare metal, the
+balance can be different - creating a need to find alternative ways to apply
+configuration changes for some scenarios.
+
+### Goals
+
+Provide a mechanism for avoiding a full cluster reboot in domain specific
+scenarios where engineering has verified that the configuration change:
+
+1. can be proven to be safely and immediately applied by some other mechanism,
+1. is not combined with other changes that do require a reboot
+1. has a demonstrable negative user impact when applied via reboot
+
+### Non-Goals
+
+1. Removing all reboots from the system
+1. Allowing the admin to configure additional domains or strategies for applying
+   configuration updates
+
+## Proposal
+
+1. Add a function to MCD that compares two MachineConfigs, and outputs an
+   ordered list of known actions that are needed to apply it
```

Comment:
How does it know what changes require what actions?

Reply:
In the absence of any other information, the function would normally return `drain+reboot` if there is any difference detected.

The intention is that in very specific cases where an alternative has been proven to be safe (such as `drain + service restart` for changes that only affect ICSP), the function will modified to return a different set of actions.

Reply:
I guess the answer here would be that we're going to base this on which files changed between 2 rendered machine configs and compute the actions based on that - we can add whatever complexity we want here also. I think the concern has been captured here https://github.com/openshift/enhancements/pull/159#discussion_r437780193 (addition vs deletion mainly)

Reply:
But I mean, how does it know that modifications to file X require a reboot but modifications to file Y don't? Especially, is this something which is hardcoded into the MCO by us, or is it something configured at runtime by the administrator?

Reply:
> But I mean, how does it know that modifications to file X require a reboot but modifications to file Y don't? 

the MCO only knows that a modification to a file X "doesn't require a reboot" - anything else will fall into "reboot as always"

> Especially, is this something which is hardcoded into the MCO by us, or is it something configured at runtime by the administrator?

The aim is to avoid leaking this choice to the admin so this logic is gonna be computed by us in the MCO (and we drive that for future cases where we might want to avoid a reboot)